beesdoo_purchase/models/purchase.py

```diff
@@ -44,42 +45,49 @@ def write(self, vals):
     def action_toggle_adapt_selling_price(self):
         for order in self:
             for line in order.order_line:
-                line.adapt_selling_price ^= True
+                line.adapt_selling_price = not line.adapt_selling_price
 
     @api.multi
-    def button_confirm(self):
-        res = super(PurchaseOrder, self).button_confirm()
+    def button_adapt_price(self):
         for order in self:
-            for line in order.order_line:
+            lines = order.order_line.filtered(
+                lambda l: l.adapt_purchase_price or l.adapt_selling_price
+            )
+            for line in lines:
                 product_id = line.product_id
                 product_tmpl_id = product_id.product_tmpl_id
                 seller = product_id._select_seller(
-                    partner_id=line.order_id.partner_id,
+                    partner_id=line.partner_id,
                     quantity=line.product_qty,
-                    date=order.date_order and order.date_order.date(),
+                    date=line.order_id.date_order
+                    and line.order_id.date_order.date(),
                     uom_id=line.product_uom,
                     params={"order_id": line.order_id},
                 )
-                price = line.price_unit
-                suggested_price = (
-                    price * product_tmpl_id.uom_po_id.factor
-                ) * (1 + product_tmpl_id.categ_id.profit_margin / 100)
-                if line.adapt_purchase_price and line.adapt_selling_price:
-                    seller.price = price
-                    # will asynchronously trigger _compute_cost()
-                    # on `product.template` in `beesdoo_product`
-                    product_tmpl_id.list_price = suggested_price
-                elif line.adapt_purchase_price:
-                    seller.price = price  # see above comment
-                elif line.adapt_selling_price:
-                    product_tmpl_id.list_price = suggested_price
-        return res
+                if seller:
+                    price = line.price_unit
+                    suggested_price = (
+                        price * product_tmpl_id.uom_po_id.factor
+                    ) * (1 + product_tmpl_id.categ_id.profit_margin / 100)
+                    if line.adapt_purchase_price and line.adapt_selling_price:
+                        # will asynchronously trigger _compute_cost()
+                        # on `product.template` in `beesdoo_product`
+                        seller.price = price
+                        product_tmpl_id.list_price = suggested_price
+                    elif line.adapt_purchase_price:
+                        seller.price = price  # see above comment
+                    elif line.adapt_selling_price:
+                        product_tmpl_id.list_price = suggested_price
+                else:
+                    raise UserError(
+                        _("Cannot adapt the price of '%s'.") % product_id.name
```

Comment:
Maybe more explicit?

No seller found, Odoo cannot adapt the price of '%s'

Reply:
done :)